components/cors-mgt/org.wso2.carbon.identity.cors.mgt.core/src/main/java/org/wso2/carbon/identity/cors/mgt/core/dao/impl/CacheBackedCORSOriginDAO.java

```diff
@@ -0,0 +1,245 @@
+import org.wso2.carbon.identity.cors.mgt.core.model.CORSOrigin;
+
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Cached DAO layer for CORS Origins. All the DAO access has to be happen through this layer to ensure
+ * single point of caching.
+ */
+public class CacheBackedCORSOriginDAO extends CORSOriginDAOImpl {
+
+    private static final Log log = LogFactory.getLog(CacheBackedCORSOriginDAO.class);
+
+    private final CORSOriginDAO corsOriginDAO;
+
+    public CacheBackedCORSOriginDAO(CORSOriginDAO corsOriginDAO) {
+
+        this.corsOriginDAO = corsOriginDAO;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int getPriority() {
+
+        return 5;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<CORSOrigin> getCORSOriginsByTenantId(int tenantId) throws CORSManagementServiceServerException {
+
+        CORSOrigin[] cachedResult = getCORSOriginsFromCache(tenantId);
+        if (cachedResult != null) {
+            return Arrays.asList(cachedResult);
+        }
+
+        List<CORSOrigin> corsOrigins = corsOriginDAO.getCORSOriginsByTenantId(tenantId);
+        addCORSOriginsToCache(corsOrigins.toArray(new CORSOrigin[0]), tenantId);
+        return corsOrigins;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<CORSOrigin> getCORSOriginsByApplicationId(int applicationId, int tenantId)
+            throws CORSManagementServiceServerException {
+
+        CORSOrigin[] cachedResult = getCORSOriginsFromCache(applicationId, tenantId);
+        if (cachedResult != null) {
+            return Arrays.asList(cachedResult);
+        }
+
+        List<CORSOrigin> corsOrigins = corsOriginDAO.getCORSOriginsByApplicationId(applicationId, tenantId);
+        addCORSOriginsToCache(corsOrigins.toArray(new CORSOrigin[0]), applicationId, tenantId);
+        return corsOrigins;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setCORSOrigins(int applicationId, List<CORSOrigin> corsOrigins, int tenantId)
+            throws CORSManagementServiceServerException {
+
+        clearCaches(applicationId, tenantId);
+        corsOriginDAO.setCORSOrigins(applicationId, corsOrigins, tenantId);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void addCORSOrigins(int applicationId, List<CORSOrigin> corsOrigins, int tenantId)
+            throws CORSManagementServiceServerException {
+
+        clearCaches(applicationId, tenantId);
+        corsOriginDAO.addCORSOrigins(applicationId, corsOrigins, tenantId);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void deleteCORSOrigins(int applicationId, List<String> corsOriginIds, int tenantId)
+            throws CORSManagementServiceServerException {
+
+        clearCaches(applicationId, tenantId);
+        corsOriginDAO.deleteCORSOrigins(applicationId, corsOriginIds, tenantId);
+    }
+
+    /**
+     * Add CORS origins to the cache.
```

Comment:
Can we improve the method comment to specifically describe the functionality of the method. Since this method only requires `tenantId`? Update the other `addCORSOriginsToCache` method comment also.

Reply:
Fixed